Add first pass at AWS Lambda background job documentation (LG-3676) (#45)
18F/identity-handbook: _articles/appdev-lambda-jobs.md

      |                      |                      |
      v           +--------->+ AWS Lambda Functions |
+-----+-----+     |          |                      |
|           +-----+          ++---------+-----------+
|    IDP    |                 |         |
|           |    HTTP POST    |         v
|           +<----------------+   Outbound Proxy
+--+-----+--+                           |
   |     |                              |
   |     |                              |
   |     |                              v
   |     |                           Vendors
   v     v
Redis  Postgres
```
(to update this diagram, copy-paste this to and from <http://asciiflow.com>)

The lifecycle of a lambda:

1. (For image processing jobs only)
  - The IDP will generate pre-signed S3 URLs and pass them to the client/browser
  - The browser will generate random encryption keys
  - The browser will AES-encrypt images and upload those encrypted images to the pre-signed S3 URLs
2. The user submits a form to the IDP
    - For image processing jobs, the payload will contain:
        - S3 image URLs
        - the encryption keys and IV
    - For PII verification jobs, the payload will contain PII:

>         - First name

Reviewer: I think at least some of the address-based ones are sent too, and as far as I know any PII received from doc auth is included too.

Author: address added in 987a71c851294faf09ba529d60797931721c152b

    - The lambda payload contains a `callback_url` which has a unique token for the job
    - The lambda payload contains the values submitted from the form (S3 image urls, encryption keys, or PII as needed)
4. The IDP will show a waiting page, refreshed via `<meta>` refresh tag
5. The Lambda will process the jobs
    - (For image processing jobs only)
        - The Lambda will download the encrypted image data from S3, and decrypt it
    - It will make HTTP requests via our outbound proxy to vendors
        - The request to the vendors will include either PII or decrypted image data
    - When the lambda is done, it will HTTP POST back to the IDP
        - It uses the `callback_url` with a unique token
        - It authenticates by passing a `X-API-AUTH-TOKEN` header (shared secret) with the IDP, the token
          is different per job (3 different jobs means there are 3 different authentication tokens)
        - The IDP stores the result (which may contain PII) in Redis, symmetrically encrypted and
          with a 60 second expiration
7. On next `<meta>` refresh, the IDP will check Redis for the result for that particular job,
   and then continue with the rest of the flow

> 

Reviewer: Should a step related to timeouts be included?

Author: Good point! added in b599dbb2bae2f75462c7b0fd21e0872743c114c5

[git-ref]: https://github.com/18F/identity-idp/blob/master/app/services/lambda_jobs/git_ref.rb

## Deploys

The code that runs inside the AWS Lambdas lives in the [identity-idp-functions][identity-idp-functions] repository.

We use CircleCI ([`.circleci/config.yml`][circleci]) to build and package the code via `sam`, and upload it to S3. The uploads
are tagged by their git ref.

> TODO: add more info about the codebuild pipeline after the files get to S3.

Author: @crissupb I'd love to get a better understanding of the deploy side, can you help me add a high-level description and maybe some links to our other repos where things get configured?

Author: (gonna merge this now and we can follow-up with more of this part later)

Generally, the jobs fall into two categories:

1. ID card scanning (image processing jobs)
2. PII verification

```
      +--------->  AWS S3  <------------+
      |                                 |
      |                                 |
 User (browser)                         |
      |                                 |
      |                      +----------+-----------+
      |                      |                      |
      v           +--------->+ AWS Lambda Functions |
+-----+-----+     |          |                      |
|           +-----+          ++---------+-----------+
|    IDP    |                 |         |
|           |    HTTP POST    |         v
|           +<----------------+   Outbound Proxy
+--+-----+--+                           |
   |     |                              |
   |     |                              |
   |     |                              v
   |     |                           Vendors
   v     v
Redis  Postgres
```
(to update this diagram, copy-paste this to and from <http://asciiflow.com>)

The lifecycle of a lambda:

1. (For image processing jobs only)
  - The IDP will generate pre-signed S3 URLs and pass them to the client/browser

>   - The browser will generate random encryption keys
>   - The browser will AES-encrypt images and upload those encrypted images to the pre-signed S3 URLs
> 2. The user submits a form to the IDP
>     - For image processing jobs, the payload will contain:
>         - S3 image URLs
>         - the encryption keys and IV

Reviewer 2: Part minor correction, part self-validation: We only generate a single key, but unique IVs (per field):

```suggestion
  - The browser will generate a random encryption key
  - The browser will AES-encrypt images and upload those encrypted images to the pre-signed S3 URLs
2. The user submits a form to the IDP
    - For image processing jobs, the payload will contain:
        - S3 image URLs
        - the encryption key and IVs
```

Author: Sounds good, thanks!

    - For PII verification jobs, the payload will contain PII:
        - First name
        - Last name
        - Date of Birth
        - SSN
        - Driver's license number
        - Address
3. The IDP immediately kicks off a Lambda (IDP does not persist any of the form data)
    - The IDP is pinned to a particular SHA of the Lambda code, via its its source (see [`GIT_REF`][git-ref] constant)
    - The lambda payload contains a `callback_url` which has a unique token for the job
    - The lambda payload contains the values submitted from the form (S3 image urls, encryption keys, or PII as needed)

> 4. The IDP will show a waiting page, refreshed via `<meta>` refresh tag

Reviewer 2: Not sure if this is something that's planned to change at all pending upcoming design revisions? In any case, it varies by job type as well, where the image processing (document capture) doesn't use `meta` tag. Maybe leave out that implementation detail?

```suggestion
4. The IDP will show a waiting page to the user
```

Author: Yeah, makes sense to keep this part generic

5. The Lambda will process the jobs
    - (For image processing jobs only)
        - The Lambda will download the encrypted image data from S3, and decrypt it
    - It will make HTTP requests via our outbound proxy to vendors
        - The request to the vendors will include either PII or decrypted image data
6. When the lambda is done, it will HTTP POST back to the IDP
    - It uses the `callback_url` with a unique token
    - It authenticates by passing a `X-API-AUTH-TOKEN` header (shared secret) with the IDP, the token
      is different per job (3 different jobs means there are 3 different authentication tokens)
    - The IDP stores the result (which may contain PII) in Redis, symmetrically encrypted and
      with a 60 second expiration.
    - PII in the payload may include data from reading the driver's licese
        - First name
        - Last name
        - Date of Birth
        - Driver's license number
        - Address

> 7. On next `<meta>` refresh, the IDP will check Redis for the result for that particular job,
>    and then continue with the rest of the flow

Reviewer 2: Same as above, though maybe we can make it about "polling" broadly? Whether that's `meta` or otherwise.

```suggestion
7. The user waiting page will be polling for the result of the background job, where the IDP will
   check Redis for the result for that particular job. Once it is complete, the user will continue
   to the next step of the flow.
```
